src/opustrainer/trainer.py

```diff
@@ -37,12 +136,16 @@ def apply_titlecase(line: str) -> str:
         sections[i] = ' '.join([word[0].upper() + word[1:] for word in sections[i].split()])
     return '\t'.join(sections)
 
+class ModifierType(Enum):
+    BATCH = 0
+    SENTENCE = 1
+
 MODIFIERS = {
-    'uppercase': lambda line: line.upper(),
-    'titlecase': apply_titlecase,
+    'uppercase': (ModifierType.BATCH, lambda line: line.upper()),
+    'titlecase': (ModifierType.BATCH, apply_titlecase),
+    'tags': (ModifierType.SENTENCE, tag_line)
 }
```

Comment:
I'm tempted to rework this at some point into a generic class for each type, that also supports instantiation so we only load and initialise the moses tokenizer when the tags modifier is applied.

I'm still a bit confused by how `probability` for BATCH is per-sentence probability, but in SENTENCE it means per-word probability. My main issue being that you cannot create a scenario now where there are few sentences with the tag_line filter applied, but in those lines there are almost always multiple tags. Right now, I'm assuming you set the probabilty quite low for it to not have all input sentences have a placeholder. But then I'd also expect the number of sentences with 3 or more placeholders to be really, really low. (The fact that you have a working system is kind of proof that my worries are only theoretical 😅 )

Reply:
I am not sure what your issue is here other than the names of the BATCH and SENTENCE being confusing (Maybe they should be SENTENCE and WORD instead?
I currently set the probability to 0.02 or 0.01 and that is applied to every single word in every single sentence. In practice it doesn't overwhelm the training with bogus examples.

Reply:
My point about having just a WORD probability is that even with 0.02 you change 1/3 of all your sentences, but you'll almost never see 3 or more mutations in a single sentence. That's why I'd advocate for a SENTENCE + a WORD probability.

```py
import numpy as np
import matplotlib.pyplot as plt
from random import random, randint

PROBABILITY = 0.02

# lets take 1000 sentences (just storing their length here)
sentences = (
	randint(3, 50)
	for _ in range(1_000_000)
)

mutations = [
    sum(1 if random() < PROBABILITY else 0 for _ in range(length))
    for length in sentences
]

plt.hist(mutations, bins=list(range(0,8)))
```
![image](https://user-images.githubusercontent.com/198639/226685960-8bf90685-28e1-420e-a04d-609d9dbf4742.png)

Reply:
So, to summarize what we spoke about today:
1) We only consider tag candidates to be one-to-one alignment mappings which are very much not every single word in the sentence.
2) We ignore one-to-one alignments that are the same on both the source and the target (numbers, punctuation, etc).
3) Since each tag in a sentence is unique, we expect that we would be able to handle a sentence that has multiple tagged words without necessarily seeing too many of them during training.

All that being said,  word + sentence probability would just be another more granular way to define those.

Reply:
Renamed to SENTENCE and WORD respectively